File: viewer/src/state/session.ts
import { create } from "zustand";

export interface PoseSample {
  t: number;
  p: [number, number, number];
  q: [number, number, number, number];
  tracking: "normal" | "limited" | "notAvailable";
  trackingReason?:
    | "initializing"
    | "relocalizing"
    | "excessiveMotion"
    | "insufficientFeatures";
}

export interface ImuSample {
  t: number;
  a: number[];
  ua: number[];
  rr: number[];
  qi: number[];
  mag?: number[];
}

export interface TouchEvent {
  t: number;
  kind: "start" | "move" | "end";
  id: number;
  x: number;
  y: number;
  force?: number;
}

export interface Marker {
  t: number;
  label: string;
}

export interface Session {
  version: 1;
  id: string;
  createdAt: string;
  device: {
    model: string;
    iosVersion: string;
    hasLiDAR: boolean;
    screen: { width: number; height: number; scale: number };
  };
  pose: PoseSample[];
  imu: ImuSample[];
  touches: TouchEvent[];
  markers: Marker[];
  duration: number;
}

export interface LiveBatch {
  sessionId?: string;
  pose?: PoseSample[];
  imu?: ImuSample[];
  touches?: TouchEvent[];
}

interface SessionState {
  mode: "live" | "replay";
  setMode: (mode: "live" | "replay") => void;

  liveSessionId: string | null;
  livePoses: PoseSample[];
  liveTouches: TouchEvent[];
  pushLiveBatch: (batch: LiveBatch) => void;

  replaySession: Session | null;
  setReplaySession: (session: Session | null) => void;

  playbackT: number;
  setPlaybackT: (t: number) => void;

  playing: boolean;
  setPlaying: (v: boolean) => void;

  speed: number;
  setSpeed: (v: number) => void;

  cameraFollow: boolean;
  setCameraFollow: (v: boolean) => void;
}

const MAX_LIVE_POSES = 2000;
const MAX_LIVE_TOUCHES = 200;

export const useSessionStore = create<SessionState>((set) => ({
  mode: "live",
  setMode: (mode) => set({ mode }),

  liveSessionId: null,
  livePoses: [],
  liveTouches: [],
  pushLiveBatch: (batch) =>
    set((state) => {
      const nextPoses = batch.pose
        ? [...state.livePoses, ...batch.pose].slice(-MAX_LIVE_POSES)
        : state.livePoses;
      const nextTouches = batch.touches
        ? [...state.liveTouches, ...batch.touches].slice(-MAX_LIVE_TOUCHES)
        : state.liveTouches;
      return {
        liveSessionId: batch.sessionId ?? state.liveSessionId,
        livePoses: nextPoses,
        liveTouches: nextTouches,
      };
    }),

  replaySession: null,
  setReplaySession: (session) =>
    set({ replaySession: session, playbackT: 0, playing: false }),

  playbackT: 0,
  setPlaybackT: (t) => set({ playbackT: t }),

  playing: false,
  setPlaying: (v) => set({ playing: v }),

  speed: 1,
  setSpeed: (v) => set({ speed: v }),

  cameraFollow: false,
  setCameraFollow: (v) => set({ cameraFollow: v }),
}));

File: ios/MocapCapture/Models/Session.swift
import Foundation

// MARK: - Top-level session

struct Session: Codable {
    let version: Int
    let id: String
    let createdAt: String
    let device: DeviceInfo
    var pose: [PoseSample]
    var imu: [ImuSample]
    var touches: [TouchEvent]
    var markers: [Marker]
    var video: VideoInfo?
    var mesh: MeshInfo?
    var duration: Double   // ms
}

// MARK: - Device

struct DeviceInfo: Codable {
    let model: String
    let iosVersion: String
    let hasLiDAR: Bool
    let screen: ScreenInfo

    struct ScreenInfo: Codable {
        let width: Double
        let height: Double
        let scale: Double
    }
}

// MARK: - Pose

struct PoseSample: Codable {
    let t: Double           // ms from session start
    let p: [Double]         // [x, y, z] meters, ARKit frame
    let q: [Double]         // [x, y, z, w]
    let tracking: TrackingState
    let trackingReason: TrackingReason?

    enum TrackingState: String, Codable {
        case normal
        case limited
        case notAvailable
    }

    enum TrackingReason: String, Codable {
        case initializing
        case relocalizing
        case excessiveMotion
        case insufficientFeatures
    }
}

// MARK: - IMU

struct ImuSample: Codable {
    let t: Double           // ms from session start
    let a: [Double]         // full accel incl. gravity, m/s²
    let ua: [Double]        // user accel (gravity removed), m/s²
    let rr: [Double]        // rotation rate, rad/s
    let qi: [Double]        // attitude quaternion [x, y, z, w]
    let mag: [Double]?      // magnetometer μT (optional)
}

// MARK: - Touch

struct TouchEvent: Codable {
    let t: Double
    let kind: TouchKind
    let id: Int
    let x: Double           // 0..1 normalized
    let y: Double           // 0..1 normalized
    let force: Double?

    enum TouchKind: String, Codable {
        case start
        case move
        case end
    }
}

// MARK: - Marker

struct Marker: Codable {
    let t: Double
    let label: String
}

// MARK: - Optional media

struct VideoInfo: Codable {
    let filename: String
    let startOffsetMs: Double
    let fps: Double
    let codec: VideoCodec
    let width: Int
    let height: Int

    enum VideoCodec: String, Codable {
        case h264
        case hevc
    }
}

struct MeshInfo: Codable {
    let filename: String
}
